streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/Graph.java

```diff
@@ -0,0 +1,348 @@
+         Check provided flow satisfying below constraints:
+         1. Input flow and output flow for each node should be the same except for source and destination node
+         2. Output flow of source and input flow of destination should be the same
+        */
+
+        final Map<V, Long> inFlow = new HashMap<>();
+        final Map<V, Long> outFlow = new HashMap<>();
+        populateInOutFlow(inFlow, outFlow);
+
+        for (final Entry<V, Long> in : inFlow.entrySet()) {
+            if (in.getKey().equals(sourceNode) || in.getKey().equals(sinkNode)) {
+                continue;
+            }
+            final Long out = outFlow.get(in.getKey());
+            if (!Objects.equals(in.getValue(), out)) {
+                throw new IllegalStateException("Input flow for node " + in.getKey() + " is " +
+                    in.getValue() + " which doesn't match output flow " + out);
+            }
+        }
+
+        final Long sourceOutput = outFlow.get(sourceNode);
+        final Long sinkInput = inFlow.get(sinkNode);
+        if (!Objects.equals(sourceOutput, sinkInput)) {
+            throw new IllegalStateException("Output flow for source " + sourceNode + " is " + sourceOutput
+                + " which doesn't match input flow " + sinkInput + " for sink " + sinkNode);
+        }
+    }
+
+    private void cancelNegativeCycles() {
+        if (!isResidualGraph) {
+            throw new IllegalStateException("Should be residual graph to cancel negative cycles");
+        }
+        boolean cyclePossible = true;
+        while (cyclePossible) {
+            cyclePossible = false;
+            for (final V node : nodes) {
+                final Map<V, V> parentNodes = new HashMap<>();
+                final Map<V, Edge> parentEdges = new HashMap<>();
+                final V nodeInCycle = detectNegativeCycles(node, parentNodes, parentEdges);
+                if (nodeInCycle == null) {
+                    continue;
+                }
+                cyclePossible = true;
+                V parentNode = parentNodes.get(nodeInCycle);
+                Edge parentEdge = parentEdges.get(nodeInCycle);
+
+                // Find max possible negative flow
+                int possibleFlow = parentEdge.residualFlow;
+                for (V curNode = parentNode; curNode != nodeInCycle; curNode = parentNodes.get(curNode)) {
+                    parentEdge = parentEdges.get(curNode);
+                    possibleFlow = Math.min(possibleFlow, parentEdge.residualFlow);
+                }
+
+                // Update graph by removing negative flow
+                parentNode = parentNodes.get(nodeInCycle);
+                parentEdge = parentEdges.get(nodeInCycle);
+                Edge counterEdge = parentEdge.counterEdge;
+                parentEdge.residualFlow -= possibleFlow;
+                parentEdge.flow += possibleFlow;
+                counterEdge.residualFlow += possibleFlow;
+                if (counterEdge.flow >= possibleFlow) {
```

Comment:
Not sure if I get this condition. Can you elaborate?

Reply:
If there's some flow going in one direction, there should be the same amount of flow going in opposite direction. 

In the case of forward edge, the `counterEdge.flow` is always larger than `possibleFlow` because forward edge's flow is backward edge's residual flow and `possibleFlow` is always smaller than or equal to residual flow.

In the case of backward edge, `counterEdge.flow` can be smaller than `possibleFlow` because forwardEdge's residual flow is bounded by capacity. Actually in case's `counterEdge.flow < possibleFlow`, we also need to reset it to 0. I'll fix that.

Reply:
OK. After closer look, I think updating backward edge doesn't make much sense. So I added a `forwardEdge` bool in Edge to indicate whether it's forward edge and only update the flow in forward edge.